app/frontend/wallet/shelley/transaction/computeTxPlan.ts

```diff
@@ -0,0 +1,268 @@
+import {InternalError, InternalErrorReason, UnexpectedErrorReason} from '../../../errors'
+import {CertificateType, Lovelace} from '../../../types'
+import {MAX_TX_OUTPUT_SIZE} from '../../constants'
+import {aggregateTokenBundles, getTokenBundlesDifference} from '../../helpers/tokenFormater'
+import {TxAuxiliaryData, TxCertificate, TxInput, TxOutput, TxWithdrawal} from '../../types'
+import {cborizeSingleTxOutput} from '../shelley-transaction'
+import {TxPlanResult} from './types'
+import {
+  computeChange,
+  computeMinUTxOLovelaceAmount,
+  computeRequiredDeposit,
+  computeRequiredTxFee,
+} from './utils'
+import {encode} from 'borc'
+import {MAX_OUTPUT_TOKENS, MIN_UTXO_VALUE} from './constants'
+
+export function computeTxPlan(
+  inputs: Array<TxInput>,
+  outputs: Array<TxOutput>,
+  possibleChange: TxOutput,
+  certificates: Array<TxCertificate>,
+  withdrawals: Array<TxWithdrawal>,
+  auxiliaryData: TxAuxiliaryData
+): TxPlanResult {
+  const totalRewards = withdrawals.reduce((acc, {rewards}) => acc + rewards, 0)
+  const totalInput = inputs.reduce((acc, input) => acc + input.coins, 0) + totalRewards
+  const totalInputTokens = aggregateTokenBundles(inputs.map(({tokenBundle}) => tokenBundle))
+  const deposit = computeRequiredDeposit(certificates)
+  const totalOutput = outputs.reduce((acc, {coins}) => acc + coins, 0) + deposit
+  const totalOutputTokens = aggregateTokenBundles(outputs.map(({tokenBundle}) => tokenBundle))
+
+  // total amount of lovelace that had to be added to token-containing outputs
+  const additionalLovelaceAmount = outputs.reduce(
+    (acc, {coins, tokenBundle}) => (tokenBundle.length > 0 ? acc + coins : acc),
+    0
+  ) as Lovelace
+
+  const feeWithoutChange = computeRequiredTxFee(
+    inputs,
+    outputs,
+    certificates,
+    withdrawals,
+    auxiliaryData
+  )
+
+  const tokenDifference = getTokenBundlesDifference(totalInputTokens, totalOutputTokens)
+
+  // Cannot construct transaction plan, not enought tokens
+  if (tokenDifference.some(({quantity}) => quantity < 0)) {
+    return {
+      success: false,
+      minimalLovelaceAmount: additionalLovelaceAmount,
+      estimatedFee: feeWithoutChange,
+      deposit,
+      error: {code: UnexpectedErrorReason.CannotConstructTxPlan},
+    }
+  }
+
+  // Cannot construct transaction plan, not enough lovelace
+  if (inputs.length === 0 || totalOutput + feeWithoutChange > totalInput) {
+    return {
+      success: false,
+      minimalLovelaceAmount: additionalLovelaceAmount,
+      estimatedFee: feeWithoutChange,
+      deposit,
+      error: {code: UnexpectedErrorReason.CannotConstructTxPlan},
+    }
+  }
+
+  // No change necessary, perfect fit
+  if (totalOutput + feeWithoutChange === totalInput && tokenDifference.length === 0) {
```

Comment:
what if you are sending exactly one token out and it's a perfect match when it comes to ADA? Wouldn't the tokenDifference be non-empty, hence causing the code to produce an invalid change output with zero ADA and zero tokens?

Reply:
so perhaps the check should be

```javascript
tokenDifference.every(({quantity}) => quantity === 0)
```

Reply:
Agree, I think the situation you explained in the first comment wont happen since tokenDifference.length is 0 if its perfect fit (0 differences are omitted) but I would suggest creating a variable called `isTokenDifferenceEmpty` which would be defined as you proposed to emphasize this.